affine.scad

```diff
@@ -230,8 +235,238 @@ function rot_decode(M) =
         tproj = translation - (translation*axis)*axis,    // Translation perpendicular to axis determines centerpoint
         cp    = (tproj + cross(axis,tproj)*c_cos/c_sin)/2
     )
-    [angle, axis, cp, (translation*axis)*axis];
+    [long ? 360-angle:angle,
+     long? -axis : axis,
+     cp,
+     (translation*axis)*axis];
+
 
+// Function: rot_inverse()
+// Usage:
+//   B = rot_inverse(A)
+// Description:
+//   Inverts a 2d or 3d rotation matrix.  The matrix can be a rotation around any center,
+//   so it may include a translation.
+function rot_inverse(T) =
+    assert(is_matrix(T,square=true),"Matrix must be square")
+    let( n = len(T))
+    assert(n==3 || n==4, "Matrix must be 3x3 or 4x4")
+    let(
+        rotpart =  [for(i=[0:n-2]) [for(j=[0:n-2]) T[j][i]]],
+        transpart = [for(row=[0:n-2]) T[row][n-1]]
+    )
+    assert(approx(determinant(T),1),"Matrix is not a rotation")
+    concat(hstack(rotpart, -rotpart*transpart),[[for(i=[2:n]) 0, 1]]);
+
+
+function _closest_angle(alpha,beta) =
+    is_vector(beta) ? [for(entry=beta) _closest_angle(alpha,entry)]
+  : beta-alpha > 180 ? beta - ceil((beta-alpha-180)/360) * 360 
+  : beta-alpha < -180 ? beta + ceil((alpha-beta-180)/360) * 360
+  : beta;
+
+
+// Smooth data with N point moving average.  If angle=true handles data as angles.
+// If closed=true assumes last point is adjacent to the first one.
+// If closed=false pads data with left/right value (probably wrong behavior...should do linear interp)
+function _smooth(data,len,closed=false,angle=false) =
```

Comment:
This name seems likely to result in namespace collision with a function in an end user's code.

Reply:
Even hidden with the underscore?   Isn't our convention that we rule the underscore namespace?  

I actually wondered if that function should be exposed...but wasn't feeling like figuring it out.  You can rename it to something more verbose if it seems necessary.   It's a general moving average that handles angles, but maybe not the best with end effects.   And I think it fails if length is even.

Reply:
You have a point about the _ prefix.